### DISCUSSION TITLE
MASS (Mueen's Similarity Search) Algorithm

### DISCUSSION THREAD

**Discussion by Deepakgthomas** [TDAmeritrade/stumpy#62] [Q&A]
Hi Sean,

Does STUMPY provide Numba compatible function where the MASS matrix can be calculated?

Thanks,
Deepak Thomas

**Comment by seanlaw** [comment]
@Deepakgthomas Given the context of MPdist, this is an excellent question!

All of the basic functions such as MASS are stored in core.py or stamp.py. However, since MASS is typically only called once then I had not considered JIT compiling them with Numba since they were already NumPy functions.

Of course, all of this changes when MASS gets called many times as will be the case with MPdist, FLOSS, and STUMPI (STUMP Incremental).

Having said that, I don’t think there will be much work beyond adding a @njit decorator and seeing if things break. But let’s just get it to work first and optimize after the first PR for the tutorial. It is good to start a checklist of “minimum viable issue” and a separate list of “wish list items”. 

From my limited experience, unit testing both JIT compiled and Python native code for the same function can be tricky. We already do this now.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by seanlaw** [comment]
I am a self proclaimed perfectionist by nature and it's a bad habit. Having said that, assuming that the code is readable and all functions are tested, “perfect is the enemy of good”. So, let’s work on the tutorial first and get to good 😊

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Deepakgthomas** [comment]
Oh of course. This was for another one of my projects. I was looking for a really fast implementation of MASS.
